src/sqlfluff/core/parser/grammar/sequence.py

```diff
@@ -295,43 +310,59 @@ def match(
             pre_nc = ()
             post_nc = ()
 
-        # If we don't have anything left after trimming, act accordingly.
+        # If we've got a case of empty brackets check whether that is allowed.
         if not content_segs:
             if not self._elements or (
-                all(e.is_optional() for e in self._elements) and self.allow_gaps
+                all(e.is_optional() for e in self._elements)
+                and (self.allow_gaps or (not pre_nc and not post_nc))
```

Comment:
I should probably know this, but what's `nc`?

Reply:
ah, _non code_. I'll add some comments

Reply:
Actually.... I'll just rename them to `pre_segs` and `post_segs` 👍